Compatibility fixes and test cases

sebdeckers opened the pull request:
Accessing the request's headers was returning `undefined`. This is a quick fix that converts the headers object to a Map. Probably not the best for performance, but I wasn't sure if you'd rather rewrite all the property getters that expect `.has` and `.get` methods.

mcollina commented:
wooow, thanks! Can you add a quick unit test for this? You can see an example in: https://github.com/nodejs/http2/blob/master/test/parallel/test-http2-create-client-session.js.

I'm not 100% sure what was the problem in the first place, so the test would help as well.

sebdeckers commented:
@mcollina The problem was that `stream[kHeaders]` as always `undefined`.

Actually I don't quite like the `Map` solution. I will rewrite this with `Object.prototype.hasOwnProperty` instead, so that the headers remain a plain object.

Will also work on a test case, sure. 😄

jasnell commented:
Why not pass the `headers` object off directly so that it does not need conversion?

```js
stream[kHeaders] = headers;
```

sebdeckers commented:
@jasnell Yep, doing that now. Got confused because the property getters use get/has. Not sure why...?

sebdeckers commented:
@jasnell Pushed a new commit that uses the plain object. No more conversion.

@mcollina Added a separate test script for compat. Made it extensible so I can also write test cases for other parts of that file (i.e. response).

sebdeckers commented:
PS: I see some unfinished provisions for implementing trailers on `Http2ServerRequest`. I believe this would involve listening to the `stream.trailers` event and storing them as `this[kTrailers]`. Possibly need to validate and merge multiple trailers. Is this a WIP? Happy to take a look at implementing this.

sebdeckers commented on lib/internal/http2/core.js:
```diff
@@ -441,6 +441,8 @@ function afterDoStreamWrite(status, handle, req) {
 
 function onHandleFinish() {
   var session = this[kSession];
+  if (!session)
```
Btw this feels like a code smell. I found that `onHandleFinish` always runs twice. Second time the session is `undefined`, hence the guard.

jasnell replied:
Yep, this was an area that I've been planning to get back to (specifically, the sequencing of the events and a well defined event lifecycle)

sebdeckers commented:
@mcollina @jasnell I did some more work over the weekend to fix the compatibility layer. Please take a look and let me know what you think.

My aim is to make it work similar to Node.js' http APIs. Also added more tests, as requested. Probably covers some of #42.

- Am I going in the right direction with this PR?
- Anything I should clean up or refactor?
- How do you typically communicate on this project (Gitter, Slack, Hangouts)?

mcollina commented on lib/internal/http2/core.js:
```diff
@@ -1055,9 +1063,9 @@ function createClientSession(options, socket) {
-  session.on('selectPadding', sessionOnSelectPadding);
+  // session.on('error', sessionOnError);
+  // session.on('stream', sessionOnStream);
+  // session.on('selectPadding', sessionOnSelectPadding);
```
why those are commented?

sebdeckers replied:
These are event handlers for the server. They each use `session[kServer]` which is undefined for clients. I was wondering if we need these handlers at all, or if they should do something else perhaps. Thoughts?

jasnell replied:
These are not just for the server side. Each of these will be relevant for the client side also.

sebdeckers replied:
@jasnell These events are certainly relevant to the client as a public interface. I used the `session:stream` event in one of the tests. But these internally assigned handlers were server-specific. It is not clear to me how this should be rewritten. Suggestion?

jasnell replied:
will be thinking through this bit more now that time is freeing up to work on it again :-) hopefully more suggestions soon.

mcollina commented on lib/internal/http2/compat.js:
```diff
@@ -330,45 +306,97 @@ class Http2ServerResponse extends Stream {
+  }
+
+  writeHeader(statusCode, headers) {
+    return this.writeHead.apply(this, arguments);
```
why this change? `apply(this, arguments)` is super slow.

sebdeckers replied:
TIL `apply`/`call` are about [half the speed](https://jsperf.com/function-calls-direct-vs-apply-vs-call-vs-bind/6) of a direct call on Chrome.

Saw writeHeaders in `_http_server.js` being forwarded to writeHead [exactly this way using `apply`](https://github.com/nodejs/http2/blob/master/lib/_http_server.js#L232-L234).

Looks like [the writeHeaders API has been deprecated](https://github.com/nodejs/node/commit/fb71ba49212dd21c0404892b4bddc37437bd5397) and dropped from the main nodejs/node repo.

Will remove it.

mcollina commented on lib/internal/http2/compat.js:
```diff
@@ -330,45 +306,97 @@ class Http2ServerResponse extends Stream {
+      assertValidHeader(name, value);
+      var trailers = this[kTrailers];
+      if (!trailers)
+        trailers = this[kTrailers] = Object.create(null);
```
@jasnell can you check this? Is `Object.create(null)` ok?

sebdeckers replied:
Looking at a [jsPerf test case](https://jsperf.com/object-create-null-vs-literal/2), it seems `{}` is 10-20x faster than `Objet.create(null)`. Should change it to that, I guess.

jasnell replied:
`Object.create(null)` is better in this case.

jasnell replied:
The more I think about this, the more I'm thinking that we may want to go ahead and make the jump to use `Map`. The performance profile is significantly improved and should be better by the time this lands.

mcollina commented on lib/internal/http2/compat.js:
```diff
@@ -428,20 +456,31 @@ class Http2ServerResponse extends Stream {
     if (!state.headersSent) {
       state.headersSent = true;
-      const headers = llistToHeaders(this[kHeaders]);
+      const headers = this[kHeaders] || Object.create(null);
```
@jasnell I think `llistToHeaders` is needed, or am I missing something?

sebdeckers replied:
TIL: The [nghttp2 `nghttp2_on_header_callback` API](https://nghttp2.org/documentation/types.html#c.nghttp2_on_header_callback) fires one header (name/value pair) at a time, so it is reasonable to append decoded headers to a (linked) list as they are received. This may explain why the linked list was used on the JS side.

jasnell replied:
yes, this is exactly why a linked list was used. I'll have to dig in a bit more to see what is best here.

jasnell commented:
I'll take a more detailed look through in the morning.

sebdeckers commented:
@mcollina Thank you for the feedback! 🙇 Have now separated the tests into separate files. Also linted the code.

Notes to self:

Run Eslint as per the Makefile:
```
./node tools/eslint/bin/eslint.js --cache --rulesdir=tools/eslint-rules "test/parallel/test-http2-*"
```

Quickly run the relevant test cases in bash:
```
for test in test/parallel/test-http2-*.js; do ./node $test; done
```

sebdeckers commented:
WIP:

- Cookies are not currently supported the same way the http1 module handles them.

- createPushResponse does not yet follow the API design shown in the [implementation nodes](https://github.com/nodejs/http2/blob/master/doc/http2-implementation-notes.md#example). For now it simply returns the ServerHttp2Stream.

mcollina commented:
> Cookies are not currently supported the same way the http1 module handles them.

can you please open a separate issue about cookies?

sebdeckers commented:
In 373969563 I've implemented a minimalist API for push requests and push responses.

```js
server.on('request', (request, response) => {
  const headers = {
    ':path': '/foobar.txt',
    'content-type': 'text/plain; charset=utf8'
  }

  const pushRequest = response.createPushRequest(headers)
  // createPushRequest returns an Http2PushRequest
  // Its argument is a plain Object with the PUSH_PROMISE headers (key/value).

  // Its only method is .push() which sends the PUSH_PROMISE frame,
  // and offers a callback to deal with the push response.
  pushRequest.push((error, pushResponse) => {
    // pushResponse is just an Http2ServerResponse,
    // so it can do all the usual tricks...
    pushResponse.writeHead(200, headers)
    pushResponse.end('Hello, World!')
  })
})
```

@jasnell Would you be okay with this design replacing the one from the [implementation notes example](https://github.com/nodejs/http2/blob/76dc88b4d9a140497a58559d0882c76cc5616913/doc/http2-implementation-notes.md#example)? It omits the `.path` property and `request` callback argument. Also renamed it to `createPushRequest` since that aligns with the RFC terminology (push request -> push response).

~~@mcollina I'm struggling to test the push stream using the client API. It looks like the `PUSH_PROMISE` frame is received in `onSessionHeaders` as a `NGHTTP2_HCAT_RESPONSE`, with the stream ID of the parent stream, i.e. the ID of the stream that received the frame, instead of the promised stream ID. Either I'm missing something or this may need some work on the C++ side. I'm not so good at that yet so any help is welcome. (I manually tested the server side's `createPushRequest` code using nghttp client. Pretty sure the bug is in the client side code.)~~ Update: Fixed, see https://github.com/nodejs/http2/pull/47#issuecomment-290984833

sebdeckers commented:
@mcollina @jasnell Please review the C++ changes in https://github.com/nodejs/http2/pull/47/commits/7f8a5ebe441f66e1d7a84f05ea07404948bf3762#diff-20934af26ea78010b56b52f2fc636a18

This commit adds support for promised stream ID when receiving headers. The test case in `test-http2-compat-Http2ServerResponse-createPushRequest.js` passes, so push streams can now be used in the server and client.

However I'm seeing this error on exit:

```
$ ./node test/parallel/test-http2-compat-Http2ServerResponse-createPushRequest.js
Assertion failed: (pending_callbacks_head_ == nullptr), function Free, file ../src/node-http2-core-inl.h, line 325.
Abort trap: 6
```

Any ideas? It's been 20 years since I touched C++. This has been another adventurous weekend. 😅

*Update:* Added a workaround inn 44cc0b8. By delaying the server's `pushStream.end()`, the client reads the data before closing its stream. I think this bug should be addressed in the context of #51 with a minimal test case, rather than this PR which is focused on server changes.

sebdeckers commented:
@mcollina @jasnell Any chance to get this PR reviewed & merged/rejected? Happy to do a Hangout call and discuss ways to collaborate. I'm seeing the API changes in #51 and am happy to update this PR accordingly, either in this PR or after round 11 closes. Thanks!

jasnell commented:
This PR will need to be reworked. I've updated the master against the latest core and squashed the numerous http2 related commits to clean things up and get us ready for the next phase of cleanup and adding tests, etc.

sebdeckers commented:
@jasnell Great! I'll get to work on it. 🕺

sebdeckers commented:
@jasnell I have reworked the code according to the new client API. Also rebased/squashed and then cherry-picked the patch as one commit onto the updated master branch. Hope this is acceptable.

😫 Apologies for losing the individual commits — I struggle with advanced git-fu.

jasnell commented on doc/http2-implementation-notes.md:
```diff
@@ -50,21 +50,19 @@ const server = http2.createSecureServer(options, (req, res) => {
 
```
Likely best to leave out the changes to this file... these implementation notes are out of date and will be replaced entirely very soon by proper documentation.

jasnell commented on lib/internal/http2/compat.js:
```diff
@@ -223,43 +194,48 @@ class Http2ServerRequest extends Readable {
-    var headers = stream[kHeaders];
-    return headers.get(constants.HTTP2_HEADER_METHOD);
+    var headers = this[kHeaders];
+    if (!headers)
```
I believe core is generally moving towards explicitly checking against `=== undefined` with these types of if statement

jasnell commented on lib/internal/http2/compat.js:
```diff
@@ -279,13 +255,13 @@ class Http2ServerResponse extends Stream {
     stream.on('close', onStreamClosedResponse);
     stream.on('aborted', onAborted.bind(this));
+    stream.on('streamClosed', this[kFinish].bind(this));
+    stream.on('finish', this[kFinish].bind(this));
```
Likely better to bind the function once then attach it to both...

```js
const onfinish = this[kFinish].bind(this);
stream.on('streamClosed', onfinish);
stream.on('finish', onfinish);
```

jasnell commented on lib/internal/http2/compat.js:
```diff
@@ -330,45 +306,94 @@ class Http2ServerResponse extends Stream {
     if (code >= 100 && code < 200)
       throw new RangeError('Informational status codes cannot be used');
-    if (code < 200 || code > 999)
+    if (code < 200 || code >= 600)
```
The `999` here is intentional for compatibility with the existing core implementation.

https://github.com/nodejs/node/blob/master/lib/_http_server.js#L187

jasnell commented on lib/internal/http2/compat.js:
```diff
@@ -330,45 +306,94 @@ class Http2ServerResponse extends Stream {
   }
 
+  addTrailers(headers) {
+    for (var name of Object.keys(headers)) {
```
I need to see if there are other areas where this needs to be changed, but please use a tradition `for (var n = 0; ...` style here. They out perform the `for-of` and `for-in` loops significantly.

sebdeckers replied:
Does V8 5.8, [as tested with Chrome v58](https://jsperf.co/for-vs-foreach-vs-for-of), not eliminate the performance difference?

<img width="320" alt="chrome for/for-of benchmark" src="https://cloud.githubusercontent.com/assets/78718/25647584/bf011d3c-2ff5-11e7-974c-d44a72e9fef3.png">

I can still see a 30% advantage in microbenchmarks on Node.js (this fork) with V8 5.7 using `for` loop over `for-of`. Gist: https://gist.github.com/sebdeckers/f110f4c5f65fc70da39587cb5d0fba68

```
for x 29,955,989 ops/sec ±1.58% (81 runs sampled)
for-of x 22,832,152 ops/sec ±1.64% (78 runs sampled)
```

#### Update

Using Node.js canary (v8: 6.0.125) there's a performance regression for both loop styles. Weird.

```
for x 26,097,145 ops/sec ±1.67% (82 runs sampled)
for-of x 17,436,809 ops/sec ±1.32% (77 runs sampled)
```

mcollina replied:
can you please do this with a `for` loop?

jasnell commented on lib/internal/http2/compat.js:
```diff
@@ -486,4 +520,41 @@ function onServerStream(stream, headers, flags) {
   server.emit('request', request, response);
 }
 
+class Http2PushRequest {
```
Will have to give this some more thought. At this point, at least for the early versions of this, I'm not convinced that we should expose push support via the compat API. Still stewing tho

sebdeckers replied:
FWIW my thoughts about the push API and compat layer.

1. Both npm:http2 and npm:spdy expose a compatibility layer with added support for PUSH_PROMISE.

   Quote: https://www.npmjs.com/package/http2#api
   > The goal is the perfect API compatibility, with additional HTTP2 related extensions (like server push).

1. I renamed it to `createPushRequest` to match ["Push Request" in the RFC](http://httpwg.org/specs/rfc7540.html#PushRequests), meaning the PUSH_PROMISE frame. The HEADERS/DATA frames are collectively referred to as the "Push Response," whose API is exactly the same as a regular `Http2ServerResponse`.

1. As a user of the H2 APIs it's important for me to be able to send the push requests (PUSH_PROMISE) before the main response (HEADERS/DATA for the user's request), followed finally by the push responses (HEADERS/DATA for the PUSH_PROMISE). This is the optimal flow described in: https://en.wikipedia.org/wiki/HTTP/2_Server_Push

jasnell commented on test/parallel/test-http2-binding.js:
```diff
@@ -13,7 +13,7 @@ assert(http2.SessionShutdownWrap);
 assert.strictEqual(typeof http2.SessionShutdownWrap, 'function');
 
-const settings = http2.getDefaultSettings({});
+const settings = require('http2').getDefaultSettings({});
```
Why this change?

sebdeckers replied:
This test was failing otherwise. Now matches the code just below: https://github.com/sebdeckers/http2/blob/430f9a648163b612a2240151b41e3639c4a44958/test/parallel/test-http2-binding.js#L28

jasnell replied:
this part was fixed in a separate PR. After a rebase you should be able to drop this change.

jasnell commented on test/parallel/test-http2-compat-Http2ServerRequest-headers.js:
```diff
@@ -0,0 +1,60 @@
+'use strict';
```
btw, test filenames should always be all lower case.

sebdeckers replied:
Ah right :) I knew something looked out of place... will rename them.

jasnell commented on test/parallel/test-http2-compat-Http2ServerRequest.js:
```diff
@@ -0,0 +1,58 @@
+      ':authority': 'localhost'
+    };
+    const request = client.request(headers);
+    // request.on('response', common.mustCall(function(headers) {
```
Why are these commented out? If they are todo for later, please include a `TODO(sebdeckers):` comment with a quick explanation

sebdeckers replied:
Mea culpa. Forgot to clean up these development artefacts.

sebdeckers commented:
- Updated with the latest changes in `master`.
- Renamed test files to lowercase.
- Acceptable status code limit increased to 999 & added a test case for this.
- Explicitly check for `=== undefined` or `=== false` instead of using `if (!...)` shorthand.

jasnell commented:
Tied up in the collaborator summit today but will go back through this in more detail very soon

robertkowalski commented:
ah wow, this is why i can't access headers today when trying to port tests from http to http2!

jasnell commented:
yeah, once we get this landed headers should work as needed. I should be able to get the full review done tomorrow :-)

jasnell commented on lib/internal/http2/core.js:
```diff
@@ -848,7 +863,7 @@ class ServerHttp2Stream extends Http2Stream {
       const stream = new ServerHttp2Stream(session, ret, options);
       streams.set(ret, stream);
-      process.nextTick(callback, stream, headers, 0);
+      process.nextTick(callback, null, stream, headers, 0);
```
I assume the `null` here is to make it match the typical Node.js callback `(err, ...args)` pattern? If so, I do not believe that's necessary. The objects involved here are event emitters that will emit `error` events when errors occur.

jasnell commented on lib/internal/http2/core.js:
```diff
@@ -1202,8 +1219,8 @@ function createClientSession(options, socket) {
-  session.on('stream', sessionOnStream);
-  session.on('selectPadding', sessionOnSelectPadding);
+  // session.on('stream', sessionOnStream);
+  // session.on('selectPadding', sessionOnSelectPadding);
```
For handling these, client specific handlers can be used, similar to how a few other client specific events are handled. I'd have to take a look more in depth to see what the changes would need to be. Would you mind adding a `TODO` comment on these to provide client specific implementations?

jasnell commented on test/parallel/test-http2-compat-pushrequest.js:
```diff
@@ -0,0 +1,96 @@
+const server = h2.createServer();
+server.listen(0, common.mustCall(function() {
+  server.once('request', common.mustCall(function(request, response) {
+    assert.ok(isOdd(response.stream.id));
```
Since this is only used once, perhaps best to just inline the `response.stream.id % 2 === 1` check

jasnell commented on test/parallel/test-http2-compat-pushrequest.js:
```diff
@@ -0,0 +1,96 @@
+
+    pushRequest.push(common.mustCall(function(error, pushResponse) {
+      assert.strictEqual(error, null);
+      assert.ok(isEven(pushResponse.stream.id));
```
ditto here... the separate `isEven()` and `isOdd()` methods can just be inlined.

jasnell commented on test/parallel/test-http2-compat-pushrequest.js:
```diff
@@ -0,0 +1,96 @@
+      // pushResponse.end();
+      // response.end();
+      setTimeout(() => pushResponse.end(), 100);
+      setTimeout(() => response.end(), 200);
```
Quick reminder for me... why is this delay needed again? I may have missed an earlier explanation

sebdeckers replied:
Good catch! This bug has eluded me. Seemed like the stream was being destroyed before all data had been sent/read. Possibly related to #72 and/or #77.

mcollina replied:
can you please add a TODO with a link the above bugs?

jasnell commented on test/parallel/test-http2-compat-pushrequest.js:
```diff
@@ -0,0 +1,96 @@
+      ':path': '/',
+      ':method': 'GET',
+      ':scheme': 'http',
+      ':authority': 'localhost'
```
`:authority` should include the port also... so `':authority': `localhost:${server.address().port}` here

jasnell commented on test/parallel/test-http2-compat-pushrequest.js:
```diff
@@ -0,0 +1,96 @@
+      assert.strictEqual(headers[':path'], '/pushed');
+      assert.strictEqual(headers[':method'], 'GET');
+      assert.strictEqual(headers[':scheme'], 'http');
+      assert.strictEqual(headers[':authority'], 'localhost');
```
just as an aside, these standard header names are exported as constants

sebdeckers replied:
Ah yes, I was wondering why the constants are preferred. What's the difference?

jasnell commented on test/parallel/test-http2-compat-serverrequest.js:
```diff
@@ -0,0 +1,45 @@
+    };
+    const request = client.request(headers);
+    request.on('end', common.mustCall(function() {
+      client.socket.destroy();
```
Note: Once https://github.com/nodejs/http2/pull/64 lands, this can be simplified to just `client.destroy()`

sebdeckers commented:
- Rebased on master
- Added `TODO(jasnell)` note for client event handlers (core)
- Inlined `isOdd`/`isEven` (test)
- Include port in request `:authority` header (tests)
- Made `state.headersSent === true|false` checks explicit, caught a bug (compat)
- Fixed inverted headersSent check in `Http2PushRequest.push()` (compat)
- Simplified cleanup to `client.destroy()` (tests)

jasnell commented on test/parallel/test-http2-compat-pushrequest.js:
```diff
@@ -0,0 +1,88 @@
+      assert.strictEqual(headers[':path'], '/pushed');
+      assert.strictEqual(headers[':method'], 'GET');
+      assert.strictEqual(headers[':scheme'], 'http');
+      assert.strictEqual(headers[':authority'], `localhost:${server.address().port}`);
```
nit: long line here.

sebdeckers commented:
Thanks for the thorough reviews, @jasnell. ☺️ I've cleaned up the linting issue and rebased once more.

mcollina commented on lib/internal/http2/compat.js:
```diff
@@ -497,4 +522,37 @@ function onServerStream(stream, headers, flags) {
+      callback(error);
+    }
+  }
+}
```
Would it be better to put all the logic in `createPushRequest()`?  Why do we need this `class`?

I promose we move the API from:

```js
const headers = {
   ':path': '/pushed',
   ':method': 'GET',
   ':scheme': 'http',
   ':authority': `localhost:${port}`
 };
 
const pushRequest = response.createPushRequest(headers);

pushRequest.push(common.mustCall(function(error, pushResponse) {
}))
```

to:

```js
const headers = {
   ':path': '/pushed',
   ':method': 'GET',
   ':scheme': 'http',
   ':authority': `localhost:${port}`
 };
 
response.createPushRequest(headers, common.mustCall(function(error, pushResponse) {
}))
```

sebdeckers replied:
Agreed. I don't know what I was thinking. 🤦‍♂️

c6a2fe2 implements the API as requested. Also renamed to `createPushResponse` since the returned object is an `Http2ServerResponse`.

mcollina commented on lib/internal/http2/compat.js:
```diff
@@ -439,20 +459,37 @@ class Http2ServerResponse extends Stream {
+      });
+    } catch (error) {
+      callback(error);
+    }
```
why are we catching this?

sebdeckers replied:
Habit from projects with other coding styles. Changed it up to simply throw instead.

mcollina commented:
@sebdeckers sorry this conflicted with @yosuke-furukawa change. Would any of you fix things?

sebdeckers commented:
@yosuke-furukawa Does the change in 6765d00 work for you? (See also [`onSessionHeaders`]( https://github.com/nodejs/http2/pull/47/commits/ed77c36ef6ac9679677c8e41de8a7b9b8ebea679#diff-696b2cc418addca5f3fe5020058f8b15R93))

The `stream` event is emitted by both server and client for any incoming streams. E.g. server receives a request, or client receives a push request. This is where request headers can be handled.

The `push` event is now emitted on the client's push request stream. This allows matching of the push response to the corresponding push request.

mcollina commented:
Landed as cee5479

jasnell commented:
@sebdeckers ... thank you for bearing with us on this. I know it took a while to get landed.

sebdeckers commented:
@jasnell I learned a lot by doing this and look forward to making more contributions. My heartfelt thanks to you and @mcollina for all the patience & guidance. 🥂
